File: packages/flutter_shared/lib/tools/investment_tool.dart
/// 复利 / 投资回报计算
class InvestmentTool {
  InvestmentTool._();

  /// 复利终值 FV = PV × (1 + r)^n
  static double futureValue(double pv, double annualRate, int years, {int compoundPerYear = 12}) {
    final r = annualRate / 100 / compoundPerYear;
    final n = years * compoundPerYear;
    return pv * _pow(1 + r, n);
  }

  /// 每月定投终值 FV = PMT × ((1 + r)^n - 1) / r
  static double futureValueWithMonthly(double pv, double monthlyInvest, double annualRate, int years) {
    final r = annualRate / 100 / 12;
    final n = years * 12;
    final fvPV = pv * _pow(1 + r, n);
    final fvPMT = monthlyInvest * (_pow(1 + r, n) - 1) / r;
    return fvPV + fvPMT;
  }

  /// 所需本金
  static double requiredPrincipal(double targetFV, double annualRate, int years) {
    final r = annualRate / 100 / 12;
    final n = years * 12;
    return targetFV / _pow(1 + r, n);
  }

  static double _pow(double x, int n) {
    double result = 1;
    for (int i = 0; i < n; i++) { result *= x; }
    return result;
  }

  static String format(double value) {
    if (value.isNaN || value.isInfinite) return '—';
    return '¥ ${value.toStringAsFixed(2)}';
  }
}

File: miniapp/src/utils/tools/investmentTool.ts
/**
 * 投资计算 — 复利 / 定投
 * 移植自 flutter_shared/tools/investment_tool.dart
 */

function pow(x: number, n: number): number {
  let r = 1;
  for (let i = 0; i < n; i++) r *= x;
  return r;
}

export const InvestmentTool = {
  /** 复利终值 */
  futureValue(pv: number, annualRate: number, years: number, compoundPerYear = 12): number {
    const r = annualRate / 100 / compoundPerYear;
    const n = years * compoundPerYear;
    return pv * pow(1 + r, n);
  },

  /** 每月定投终值 */
  futureValueWithMonthly(pv: number, monthlyInvest: number, annualRate: number, years: number): number {
    const r = annualRate / 100 / 12;
    const n = years * 12;
    const fvPV = pv * pow(1 + r, n);
    const fvPMT = monthlyInvest * (pow(1 + r, n) - 1) / r;
    return fvPV + fvPMT;
  },

  /** 所需本金 */
  requiredPrincipal(targetFV: number, annualRate: number, years: number): number {
    const r = annualRate / 100 / 12;
    const n = years * 12;
    return targetFV / pow(1 + r, n);
  },

  format(value: number): string {
    if (isNaN(value) || !isFinite(value)) return '—';
    return `¥ ${value.toFixed(2)}`;
  },
};
